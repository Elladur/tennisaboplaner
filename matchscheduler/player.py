"""Player class for the match scheduler."""

from datetime import date
from itertools import combinations


class Player:
    """A player of the match scheduler."""

    name: str
    cannot_play: list[date]

    def __init__(self, name: str, cannot_play: list[str]):
        self.name = name
        self.cannot_play = [date.fromisoformat(i) for i in cannot_play]

    def to_dict(self) -> dict:
        return {"name": self.name, "cannot_play": [str(d) for d in self.cannot_play]}

    @classmethod
    def from_dict(cls, data: dict) -> "Player":
        """Create a Player from a dictionary."""
        name = data["name"]
        cannot_play = data["cannot_play"]
        return cls(name, cannot_play)

    @classmethod
    def get_all_possible_combinations(cls, players: set["Player"]) -> list[set["Player"]]:
        """Get all possible combinations of players."""
        return [set(i) for i in combinations(players, 2)]

    def __eq__(self, __value: object) -> bool:
        if (
            __value.name == self.name  # type: ignore
            and __value.cannot_play == self.cannot_play  # type: ignore
        ):
            return True
        return False

    def __str__(self) -> str:
        return self.name

    def __hash__(self) -> int:
        return hash((self.name, tuple(self.cannot_play)))

    @classmethod
    def to_string(cls, players: set["Player"]) -> str:
        """Convert a set of players to a string."""
        return ", ".join([p.name for p in players])

    @classmethod
    def set_to_tuple(cls, players: set["Player"]) -> tuple["Player", "Player"]:
        """Convert a set of players to a tuple."""
        if len(players) != 2:
            raise ValueError("A set of players must have two players.")
        return tuple(players)  # type: ignore
