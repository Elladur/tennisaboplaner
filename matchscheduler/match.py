"""Match class and factory for generating valid matches."""

from typing import Tuple

from line_profiler import profile

from .player import Player

Match = Tuple[int, int]


@profile
def create_match(player_id1: int, player_id2: int) -> Match:
    # always use smaller int in beginning
    if player_id1 < player_id2:
        return (player_id1, player_id2)
    else:
        return (player_id2, player_id1)


@profile
def can_match_be_added(rounds: list[Match], match: Match) -> bool:
    return not any([p in r for p in match for r in rounds])
    # for r in rounds:
    # if r & match != 0:
    # return False
    # return True


@profile
def get_players_of_match(match: Match) -> Match:
    return match
    # return [i for i, bit in enumerate(bin(match)[:1:-1]) if bit == '1']
    # positions = []
    # position = 0
    # while match > 0:
    # if match & 1:
    # positions.append(position)
    # match >>= 1
    # position += 1
    # return positions


def convert_match_to_string(match: int, players: list[Player]):
    return f"{players[match[0]]} vs {players[match[1]]}"
