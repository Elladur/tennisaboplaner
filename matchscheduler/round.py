"""A round of a season consisting of a list of matches."""

from line_profiler import profile

from .match import Match, convert_match_to_string, get_players_of_match
from .player import Player


@profile
def get_players_of_round(round: list[Match]) -> set[int]:
    players = []
    for m in round:
        players += list(get_players_of_match(m))
    return set(players)


def convert_round_to_string(round: list[int], players: list[Player]):
    "\n".join([convert_match_to_string(m, players) for m in round])
