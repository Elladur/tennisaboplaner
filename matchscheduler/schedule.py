"""a schedule class consisting of a list of rounds
a schedule is valid if all rounds are valid"""

from datetime import date

from .match import Match
from .player import Player
from .round import NotValidRoundError, NotValidSwapError, Round, RoundFactory


class Schedule:
    """A schedule of rounds."""

    def __init__(self, rounds: list[Round], players: set[Player]) -> None:
        self.rounds = rounds
        self.players = players

    def to_dict(self) -> dict:
        return {
            "rounds": [r.to_dict() for r in self.rounds],
            "players": [p.to_dict() for p in self.players],
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Schedule":
        rounds = []
        players = set()
        for x in data["rounds"]:
            rounds.append(Round.from_dict(x))
        for x in data["players"]:
            players.add(Player.from_dict(x))
        return cls(rounds, players)

    def is_valid(self) -> bool:
        """Check if this schedule is valid."""
        for r in self.rounds:
            if not r.is_valid():
                return False
        return True

    def get_matches(self) -> list[Match]:
        """Get a list of all matches in this schedule."""
        return [m for r in self.rounds for m in r.matches]

    def get_matches_of_player(self, player: Player) -> list[Match]:
        """Get a list of all matches in this schedule with the given player."""
        return [m for m in self.get_matches() if player in m.players]

    def get_matches_of_players(self, players: set[Player]) -> list[Match]:
        """Get a list of all matches in this schedule with the given player."""
        return [m for m in self.get_matches() if players == m.players]

    def swap_matches(
        self, round_index1: int, match_index1: int, round_index2: int, match_index2: int
    ) -> None:
        """Swap the matches in the given rounds and match indices."""
        round1 = self.rounds[round_index1]
        round2 = self.rounds[round_index2]
        match1 = round1.matches[match_index1]
        match2 = round2.matches[match_index2]
        try:
            round1.swap_players(match_index1, match2.players)
            round2.swap_players(match_index2, match1.players)
        except NotValidSwapError as exc:
            # revert swap
            round1.swap_players(match_index1, match1.players)
            round2.swap_players(match_index2, match2.players)
            # reraise error
            raise exc
        if not self.is_valid():
            raise NotValidScheduleError("Swap is not valid.")


class NotValidScheduleError(Exception):
    """Raised when a schedule is not valid."""


class ScheduleFactory:
    """A factory for schedules."""

    @staticmethod
    def generate_valid_schedule(
        players: set[Player], dates: list[date], num_courts: int
    ) -> Schedule:
        """Generate a valid schedule for the given players and dates."""
        rounds = []
        for d in dates:
            try:
                rounds.append(RoundFactory.generate_valid_round(players, d, num_courts))
            except NotValidRoundError as exc:
                raise NotValidScheduleError("Could not generate a valid schedule.") from exc

        s = Schedule(rounds, players)
        if s.is_valid():
            return s
        raise NotValidScheduleError("Could not generate a valid schedule.")
