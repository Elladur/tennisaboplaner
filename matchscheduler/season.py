"""A season consisting of multiple rounds by a given start and end date."""
import random
from datetime import date, datetime, time, timedelta
from itertools import combinations

from .player import Player
from .round import NotValidSwap
from .schedule import NotValidSchedule, Schedule, ScheduleFactory


class Season:
    """A season of matches."""

    def __init__(
        self,
        players: set[Player],
        start: date,
        end: date,
        number_courts: int,
        time_start: time,
        time_end: time,
        calendar_title: str = "Tennisabo",
    ):
        self.players = players
        self.start = start
        self.end = end
        self.time_start = time_start
        self.time_end = time_end
        self.num_courts = number_courts
        self.calendar_title = calendar_title
        # generate a list of all dates for the season,
        # they start at start and occur weekly until end
        self.dates = []
        d = start
        while d <= end:
            self.dates.append(d)
            d += timedelta(days=7)

        self.schedule: Schedule | None = None

    @classmethod
    def from_dict(cls, data: dict) -> "Season":
        """Create a Season from a dictionary."""
        players = {Player.from_dict(p) for p in data["players"]}
        start = date.fromisoformat(data["abo"]["start"])
        end = date.fromisoformat(data["abo"]["end"])
        time_start = time.fromisoformat(data["calendar"]["time_start"])
        time_end = time.fromisoformat(data["calendar"]["time_end"])
        number_courts = data["abo"]["number_courts"]
        calendar_title = data["calendar"]["title"]
        return cls(
            players, start, end, number_courts, time_start, time_end, calendar_title
        )

    def generate_schedule(self) -> None:
        """Generate a schedule for this season."""
        self.schedule = ScheduleFactory.generate_valid_schedule(
            self.players, self.dates, self.num_courts
        )

    def optimize_schedule_by_swapping_players(self, swaps: int) -> int:
        """Optimize the schedule by swapping players."""
        if self.schedule is None:
            raise NotValidSchedule("No schedule generated yet.")

        current_score = self.schedule.get_score()
        # switch with all possible players
        for r in self.schedule.rounds:
            print(f"{datetime.now()}: Starting new round {r.date} ...", end="\r")
            for match_index, match in enumerate(r.matches):
                # todo: this could be optimized if we only take players
                # that are not playing in other matches as well
                current_players = match.players
                for possible_pair in Player.get_all_possible_combinations(self.players):
                    if current_players != possible_pair:
                        try:
                            r.swap_players(match_index, possible_pair)
                        except NotValidSwap:
                            continue
                        new_score = self.schedule.get_score()
                        if new_score < current_score:
                            swaps += 1
                            current_players = possible_pair
                            current_score = new_score
                        else:
                            # swap back to original match
                            r.swap_players(match_index, current_players)

        current_score = self.schedule.get_score()
        # switch players between matches of a round
        for r in self.schedule.rounds:
            print(f"{datetime.now()}: Starting new round {r.date} ...", end="\r")
            # get all combinations of match indexes
            for i, j in combinations(range(len(r.matches)), 2):
                match1 = r.matches[i]
                match2 = r.matches[j]
                for player1 in match1.players:
                    for player2 in match2.players:
                        try:
                            r.swap_players_of_existing_matches(i, j, {player1, player2})
                        except NotValidSwap:
                            continue
                        new_score = self.schedule.get_score()
                        if new_score < current_score:
                            swaps += 1
                            current_score = new_score
                        else:
                            # swap back to original matches
                            r.swap_players_of_existing_matches(i, j, {player1, player2})

        return swaps

    def optimize_schedule_by_swapping_matches(self, swaps: int) -> int:
        # todo: check if we need  this anymore if we swap also player
        # between existing matches in a round
        """Optimize the schedule by swapping matches."""
        if self.schedule is None:
            raise NotValidSchedule("No schedule generated yet.")

        indizes = [
            (i, j)
            for i in range(len(self.schedule.rounds))
            for j in range(self.num_courts)
        ]

        # shuffle index to have a random factor
        # (thus start if schedule is not to optimized)
        index_combination = list(combinations(indizes, 2))
        random.shuffle(index_combination)

        current_score = self.schedule.get_score()

        for (round_index1, match_index1), (
            round_index2,
            match_index2,
        ) in index_combination:
            print(
                f"{datetime.now()}: try swapping Round {round_index1} "
                + f"Match {match_index1} with Round {round_index2} "
                + f"Match {match_index2} ...",
                end="\r",
            )
            if (
                self.schedule.rounds[round_index1].matches[match_index1].players
                == self.schedule.rounds[round_index2].matches[match_index2].players
            ):
                continue

            try:
                self.schedule.swap_matches(
                    round_index1, match_index1, round_index2, match_index2
                )
            except NotValidSwap:
                continue
            new_score = self.schedule.get_score()
            if new_score < current_score:
                swaps += 1
                current_score = new_score
            else:
                # swap back to original matches
                self.schedule.swap_matches(
                    round_index1, match_index1, round_index2, match_index2
                )
        return swaps

    def optimize_schedule(self) -> None:
        """Optimize the schedule for this season."""
        if self.schedule is None:
            self.generate_schedule()

        swaps = 0
        while True:
            print(f"{datetime.now()}: Starting new round of optimizing ...")

            print(f"{datetime.now()}: Start swapping players ...")
            swaps += self.optimize_schedule_by_swapping_players(swaps)

            print(f"{datetime.now()}: Start swapping matches ...")
            swaps += self.optimize_schedule_by_swapping_matches(swaps)

            if swaps > 0:
                print(
                    f"{datetime.now()}: Swapped {swaps} times. "
                    + "The current score is: "
                    + f"{self.schedule.get_score()}."  # type: ignore
                )
                swaps = 0
            else:
                print(f"{datetime.now()}: No more swaps feasible.")
                break

    def export_season(self, folderpath: str) -> None:
        """Export this season to an Excel file."""
        if self.schedule is None:
            raise NotValidSchedule("No schedule generated yet.")
        self.schedule.export(
            folderpath, self.calendar_title, self.time_start, self.time_end
        )
