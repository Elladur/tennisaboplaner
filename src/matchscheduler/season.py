"""A season consisting of multiple rounds by a given start and end date."""

import itertools
import logging
import random
from datetime import date, time, timedelta

from line_profiler import profile

from .match import (Match, can_match_be_added, create_match,
                    replace_player_in_match)
from .player import Player
from .round import get_players_of_round
from .schedule import Schedule


class Season:
    """A season of matches."""

    def __init__(
        self,
        players: list[Player],
        start: date,
        end: date,
        number_courts: int,
        excluded_dates: list[date],
        schedule: Schedule,
        time_start: time,
        time_end: time,
        overall_cost: float = 0,
        calendar_title: str = "Tennisabo",
    ):
        self.players = players
        self.start = start
        self.end = end
        self.time_start = time_start
        self.time_end = time_end
        self.num_courts = number_courts
        self.excluded_dates = excluded_dates
        self.schedule = schedule
        self.calendar_title = calendar_title
        self.overall_cost = overall_cost
        self.logger = logging.getLogger(__name__)

    @classmethod
    def create(cls,
        players: list[Player],
        start: date,
        end: date,
        number_courts: int,
        excluded_dates: list[str],
        time_start: time,
        time_end: time,
        overall_cost: float = 0,
        calendar_title: str = "Tennisabo",
    ) -> "Season":
        days = []
        d = start
        while d <= end:
            if d not in excluded_dates:
                days.append(d)
            d = d + timedelta(days=7)
        schedule = Schedule.create(players, days, number_courts)
        return cls(players, start, end, number_courts, excluded_dates, schedule, time_start, time_end, overall_cost, calendar_title)

 ###### neeeds to be adapted #####
    @profile
    def change_match(self, round_index: int, match_index: int, match: Match) -> bool:
        if round_index in self.fixed_rounds:
            return False
        old_match = self.schedule[round_index][match_index]
        self.schedule[round_index][match_index] = match
        if self.check_if_round_is_valid(round_index):
            return True
        self.schedule[round_index][match_index] = old_match
        return False

    @profile
    def check_if_round_is_valid(self, round_index: int) -> bool:
        players = get_players_of_round(self.schedule[round_index])
        if len(players) != self.num_courts * 2:
            return False
        match_date = self.dates[round_index]
        return not any(match_date in self.players[p].cannot_play for p in players)

    def check_schedule_is_valid(self) -> bool:
        for i in range(len(self.schedule)):
            if not self.check_if_round_is_valid(i):
                return False
        return True

    @profile
    def swap_players_of_existing_matches(self, round_index: int, p: int, q: int) -> bool:
        if round_index in self.fixed_rounds:
            return False
        for i, match in enumerate(self.schedule[round_index]):
            self.schedule[round_index][i], swapped = replace_player_in_match(match, p, q)
            if swapped:
                continue
            self.schedule[round_index][i], swapped = replace_player_in_match(match, q, p)
        return True

    @profile
    def switch_matches(self, round1: int, match1: int, round2: int, match2: int) -> bool:
        if round1 in self.fixed_rounds or round2 in self.fixed_rounds:
            return False
        self.schedule[round1][match1], self.schedule[round2][match2] = (
            self.schedule[round2][match2],
            self.schedule[round1][match1],
        )
        if self.check_if_round_is_valid(round1) and self.check_if_round_is_valid(round2):
            return True
        self.schedule[round1][match1], self.schedule[round2][match2] = (
            self.schedule[round2][match2],
            self.schedule[round1][match1],
        )
        return False

    def to_dict(self) -> dict:
        return {
            "players": [p.to_dict() for p in self.players],
            "start": str(self.start),
            "end": str(self.end),
            "number_courts": self.num_courts,
            "time_start": str(self.time_start),
            "time_end": str(self.time_end),
            "excluded_dates": [str(d) for d in self.excluded_dates],
            "overall_cost": self.overall_cost,
            "calendar_title": self.calendar_title,
            "schedule": self.schedule,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Season":
        players = [Player.from_dict(p) for p in data["players"]]
        start = date.fromisoformat(data["start"])
        end = date.fromisoformat(data["end"])
        time_start = time.fromisoformat(data["time_start"])
        time_end = time.fromisoformat(data["time_end"])
        number_courts = data["number_courts"]
        excluded_dates = data["excluded_dates"]
        calendar_title = data["calendar_title"]
        overall_cost = data["overall_cost"]
        instance = cls(
            players,
            start,
            end,
            number_courts,
            time_start,
            time_end,
            excluded_dates,
            overall_cost,
            calendar_title,
        )
        instance.schedule = [[create_match(y[0], y[1]) for y in x] for x in data["schedule"]]
        return instance

    @classmethod
    def create_from_settings(cls, data: dict) -> "Season":
        """Create a Season from a dictionary."""
        players = [Player.from_dict(p) for p in data["players"]]
        start = date.fromisoformat(data["abo"]["start"])
        end = date.fromisoformat(data["abo"]["end"])
        excluded_dates = data["abo"]["excluded_dates"]
        time_start = time.fromisoformat(data["calendar"]["time_start"])
        time_end = time.fromisoformat(data["calendar"]["time_end"])
        number_courts = data["abo"]["number_courts"]
        overall_cost = data["abo"]["overall_cost"]
        calendar_title = data["calendar"]["title"]
        return cls(
            players,
            start,
            end,
            number_courts,
            time_start,
            time_end,
            excluded_dates,
            overall_cost,
            calendar_title,
        )
