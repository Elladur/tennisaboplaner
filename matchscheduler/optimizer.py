import logging
import random
from itertools import combinations

from line_profiler import profile

from matchscheduler.season import Season

from .match import can_match_be_added, create_match, get_players_of_match
from .player import Player
from .scoring_algorithm import ScoringAlgorithm


class Optimizer:

    def __init__(self, season: Season):
        self.season = season
        self.logger = logging.getLogger(__name__)
        self.scorer = ScoringAlgorithm()

    @profile
    def optimize_schedule_by_swapping_players(self, swaps: int) -> int:
        """Optimize the schedule by swapping players."""

        current_score = self.scorer.get_score(self.season.schedule, self.season.players)
        # switch with all possible players
        for round_index, round in enumerate(self.season.schedule):
            self.logger.debug(
                f"Switching all players: Starting new round {self.season.dates[round_index]}"
            )

            for match_index, current_match in enumerate(round):
                for p, q in combinations(range(len(self.season.players)), 2):
                    possible_match = create_match(p, q)
                    if possible_match == current_match:
                        continue
                    changed = self.season.change_match(round_index, match_index, possible_match)
                    if not changed:
                        continue
                    new_score = self.scorer.get_score(self.season.schedule, self.season.players)
                    if new_score < current_score:
                        swaps += 1
                        self.logger.debug(
                            f"Switched players - old score={current_score:.2f} - new score={new_score:.2f}"
                        )
                        current_score = new_score
                        current_match = possible_match
                    else:
                        # swap back to original match
                        self.season.change_match(round_index, match_index, current_match)

        current_score = self.scorer.get_score(self.season.schedule, self.season.players)
        # switch players between matches of a round
        for round_index, round in enumerate(self.season.schedule):
            self.logger.debug(
                f"Switching players inside round: Starting new round {self.season.dates[round_index]}"
            )
            # get all combinations of match indexes
            for match1, match2 in combinations(range(self.season.num_courts), 2):
                for player1, player2 in [
                    (p1, p2)
                    for p1 in get_players_of_match(round[match1])
                    for p2 in get_players_of_match(round[match2])
                ]:
                    self.season.swap_players_of_existing_matches(round_index, player1, player2)
                    new_score = self.scorer.get_score(self.season.schedule, self.season.players)
                    if new_score < current_score:
                        swaps += 1
                        self.logger.debug(
                            f"switch players inside existing round - old score={current_score:.2f} - new score={new_score:.2f}"
                        )
                        current_score = new_score
                        break
                    else:
                        # swap back to original matches
                        self.season.swap_players_of_existing_matches(round_index, player1, player2)

        return swaps

    @profile
    def optimize_schedule_by_swapping_matches(self, swaps: int) -> int:
        """Optimize the schedule by swapping matches."""
        # cant be removed even if we swap players between existing matches
        # it gives an additional random factor to the algorithmus

        indizes = [
            (i, j) for i in range(len(self.season.schedule)) for j in range(self.season.num_courts)
        ]

        # shuffle index to have a random factor
        # (thus start if schedule is not to optimized)
        index_combination = list(combinations(indizes, 2))
        random.shuffle(index_combination)

        current_score = self.scorer.get_score(self.season.schedule, self.season.players)

        for (round_index1, match_index1), (
            round_index2,
            match_index2,
        ) in index_combination:
            self.logger.debug(
                f"try swapping Round {round_index1} "
                + f"Match {match_index1} with Round {round_index2} "
                + f"Match {match_index2}"
            )
            if (
                self.season.schedule[round_index1][match_index1]
                == self.season.schedule[round_index2][match_index2]
            ):
                continue
            switched = self.season.switch_matches(
                round_index1, match_index1, round_index2, match_index2
            )
            if not switched:
                continue
            new_score = self.scorer.get_score(self.season.schedule, self.season.players)
            if new_score < current_score:
                swaps += 1
                self.logger.debug(
                    f"switched matches - old score={current_score:.2f} - new score={new_score:.2f}"
                )
                current_score = new_score
            else:
                # swap back to original matches
                self.season.switch_matches(round_index1, match_index1, round_index2, match_index2)

        return swaps

    @profile
    def optimize_schedule(self) -> float:
        """Optimize the schedule for this season."""
        swaps = 0
        while True:
            self.logger.info("Starting new round of optimizing ...")

            self.logger.info("Start swapping players ...")
            swaps += self.optimize_schedule_by_swapping_players(swaps)

            self.logger.info("Start swapping matches ...")
            swaps += self.optimize_schedule_by_swapping_matches(swaps)

            if swaps > 0:
                self.logger.info(
                    f"Swapped {swaps} times. "
                    + "The current score is: "
                    + f"{self.scorer.get_score(self.season.schedule, self.season.players)}"  # type: ignore
                )
                swaps = 0
            else:
                self.logger.info("No more swaps feasible.")
                break

        return self.scorer.get_score(self.season.schedule, self.season.players)
