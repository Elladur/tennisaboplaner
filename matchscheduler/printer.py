from datetime import datetime
import itertools
from pathlib import Path

from icalendar import Calendar, Event
from openpyxl import Workbook

from .player import Player
from .schedule import get_match_indizes_of_player, get_match_indizes_of_match
from .season import Season
from .match import convert_match_to_string, create_match


class Printer:

    def __init__(self, season: Season):
        self.season = season

    # export schedule into a excel file with each round as a row
    # consisting of the date in the first column and in every next column a match
    def export(self, folderpath: Path) -> None:
        """Export this schedule to an Excel file."""
        excel = Workbook()
        sheet = excel.active
        sheet.title = "Schedule"  # type: ignore
        sheet.append(  # type: ignore
            ["Date"] + [f"Match {i+1}" for i in range(self.season.num_courts)]
        )
        for d in sorted(self.season.dates + self.season.excluded_dates):
            if d in self.season.excluded_dates:
                sheet.append([str(d)])  # type: ignore
            else:
                i = self.season.dates.index(d)
                sheet.append([str(self.season.dates[i])] + [convert_match_to_string(m, self.season.players) for m in self.season.schedule[i]])  # type: ignore

        # add an additional sheet to excel workbook with columns for each player
        # and their match partners
        sheet = excel.create_sheet("Partner by Player")  # type: ignore
        sheet.append(  # type: ignore
            ["Date"] + [str(p) for p in self.season.players]
        )
        for i in range(len(self.season.dates)):
            row = [str(self.season.dates[i])]
            matches = self.season.schedule[i]
            for j, p in enumerate(self.season.players):
                append_string = ""
                for m in matches:
                    if j in m:
                        opponent = m[0] if m[0] != j else m[1]
                        append_string += str(self.season.players[opponent])
                        break
                row.append(append_string)
            sheet.append(row)

        # add an additional sheet to excel workbook with columns for each possible match
        # and each row marks with an x if the match is played on that day
        sheet = excel.create_sheet("Matches Overview")  # type: ignore
        player_combinations = [(p, q) for p, q in itertools.combinations(range(len(self.season.players)), 2)]
        sheet.append(  # type: ignore
            ["Date"]
            + [
                convert_match_to_string(create_match(p, q), self.season.players)
                for p, q in player_combinations
            ]
        )
        for i, round in enumerate(self.season.schedule):
            row = [str(self.season.dates[i])]
            for p, q in player_combinations:
                append_string = ""
                for m in round:
                    if m == create_match(p, q):
                        append_string = "x"
                        break
                row.append(append_string)
            sheet.append(row)

        sheet = excel.create_sheet("Costs")
        sheet.append([""] + [str(p) for p in self.season.players])  # type: ignore
        cost_per_match = self.season.overall_cost / (len(self.season.schedule) * self.season.num_courts) / 2
        sheet.append(  # type: ignore
            ["Matches"]
            + [len(get_match_indizes_of_player(self.season.schedule, p)) for p in range(len(self.season.players))]
        )
        sheet.append(  # type: ignore
            ["Cost"]
            + [len(get_match_indizes_of_player(self.season.schedule, p)) * cost_per_match for p in range(len(self.season.players))
            ]
        )

        excel.save(folderpath / "schedule.xlsx")

        # create a calendar for each player with his matches
        for i, p in enumerate(self.season.players):
            cal = Calendar()
            cal.add("prodid", "-//MatchScheduler//MatchScheduler//EN")
            cal.add("version", "2.0")
            cal.add("name", f"MatchScheduler - {p.name}")
            cal.add("X-WR-CALNAME", f"MatchScheduler - {p.name}")
            cal.add("X-WR-TIMEZONE", "Europe/Vienna")
            cal.add("X-WR-CALDESC", f"MatchScheduler - {p.name}")
            for round_index, match_index in get_match_indizes_of_player(self.season.schedule, i):
                event = Event()
                event.add("summary", self.season.calendar_title)
                event.add("description", convert_match_to_string(self.season.schedule[round_index][match_index], self.season.players))
                event.add("dtstart", datetime.combine(self.season.dates[round_index], self.season.time_start))
                event.add("dtend", datetime.combine(self.season.dates[round_index], self.season.time_end))
                cal.add_component(event)
            with open(folderpath / f"{p.name}.ics", "wb") as f:
                f.write(cal.to_ical())
