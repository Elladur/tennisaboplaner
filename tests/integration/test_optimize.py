import json
from pathlib import Path

from matchscheduler.optimizer import Optimizer
#from matchscheduler.printer import Printer
#from matchscheduler.scoring_algorithm import ScoringAlgorithm
from matchscheduler.season import Season


def test_optimize(request):
    base_path = Path(request.path).parent
    with open(f"{base_path}/input/settings.json", "r", encoding="utf-8") as input, open(
        f"{base_path}/expected/test_optimize.json", "r", encoding="utf-8"
    ) as expected:
        # load settings.json into data object
        data = json.load(input)
        s = Season.create_from_settings(data)
        o = Optimizer(s)
        score = o.optimize_schedule()

        #expected_value = ScoringAlgorithm().get_score(
            #Season.from_dict(json.load(expected)).schedule
        #)
        #assert score <= expected_value * 1.1


#def test_exclusion(request, tmp_path):
    #base_path = Path(request.path).parent
    #with open(f"{base_path}/input/test_exclusion.json", "r", encoding="utf-8") as input:
        #s = Season.create_from_settings(json.load(input))
        #o = Optimizer(s)
        #o.optimize_schedule()
        #p = Printer(s)

        #p.export(tmp_path)
        #assert s.excluded_dates != []
