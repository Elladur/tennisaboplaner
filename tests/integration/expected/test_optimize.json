{
    "players": [
        {
            "name": "player1",
            "cannot_play": [
                "2022-09-01"
            ],
            "weight": 1
        },
        {
            "name": "player2",
            "cannot_play": [
                "2022-09-08",
                "2022-09-15"
            ],
            "weight": 1
        },
        {
            "name": "player3",
            "cannot_play": [
                "2022-09-22"
            ],
            "weight": 1
        },
        {
            "name": "player4",
            "cannot_play": [
                "2022-09-29"
            ],
            "weight": 1
        },
        {
            "name": "player5",
            "cannot_play": [],
            "weight": 1
        }
    ],
    "start": "2022-09-01",
    "end": "2023-05-01",
    "number_courts": 2,
    "time_start": "19:00:00",
    "time_end": "21:00:00",
    "excluded_dates": [],
    "overall_cost": 1000,
    "calendar_title": "Tennisabo",
    "schedule": [
        [
            [
                3,
                4
            ],
            [
                1,
                2
            ]
        ],
        [
            [
                3,
                4
            ],
            [
                0,
                2
            ]
        ],
        [
            [
                3,
                4
            ],
            [
                0,
                2
            ]
        ],
        [
            [
                1,
                3
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                0,
                2
            ],
            [
                1,
                4
            ]
        ],
        [
            [
                0,
                1
            ],
            [
                2,
                3
            ]
        ],
        [
            [
                1,
                2
            ],
            [
                0,
                3
            ]
        ],
        [
            [
                2,
                4
            ],
            [
                1,
                3
            ]
        ],
        [
            [
                0,
                4
            ],
            [
                2,
                3
            ]
        ],
        [
            [
                0,
                3
            ],
            [
                1,
                4
            ]
        ],
        [
            [
                0,
                1
            ],
            [
                3,
                4
            ]
        ],
        [
            [
                2,
                3
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                0,
                2
            ],
            [
                1,
                3
            ]
        ],
        [
            [
                2,
                4
            ],
            [
                0,
                3
            ]
        ],
        [
            [
                1,
                2
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                1,
                4
            ],
            [
                2,
                3
            ]
        ],
        [
            [
                0,
                1
            ],
            [
                3,
                4
            ]
        ],
        [
            [
                0,
                2
            ],
            [
                1,
                3
            ]
        ],
        [
            [
                0,
                3
            ],
            [
                1,
                2
            ]
        ],
        [
            [
                2,
                4
            ],
            [
                0,
                1
            ]
        ],
        [
            [
                2,
                3
            ],
            [
                1,
                4
            ]
        ],
        [
            [
                1,
                3
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                0,
                3
            ],
            [
                1,
                2
            ]
        ],
        [
            [
                0,
                1
            ],
            [
                2,
                4
            ]
        ],
        [
            [
                3,
                4
            ],
            [
                0,
                2
            ]
        ],
        [
            [
                0,
                3
            ],
            [
                1,
                4
            ]
        ],
        [
            [
                1,
                2
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                1,
                3
            ],
            [
                2,
                4
            ]
        ],
        [
            [
                2,
                3
            ],
            [
                0,
                1
            ]
        ],
        [
            [
                1,
                4
            ],
            [
                0,
                2
            ]
        ],
        [
            [
                2,
                4
            ],
            [
                0,
                3
            ]
        ],
        [
            [
                1,
                3
            ],
            [
                0,
                4
            ]
        ],
        [
            [
                2,
                4
            ],
            [
                0,
                1
            ]
        ],
        [
            [
                3,
                4
            ],
            [
                1,
                2
            ]
        ],
        [
            [
                2,
                3
            ],
            [
                1,
                4
            ]
        ]
    ]
}